private-templates-service/server/src/db/providers/InMemoryDBProvider.ts

```diff
@@ -0,0 +1,210 @@
+      console.log("fuck");
+    }
+    return this._insert(doc, this.users);
+  }
+  async insertTemplate(doc: ITemplate): Promise<void> {
+    this._setTimestamps(doc);
+    return this._insert(doc, this.templates);
+  }
+
+  async getUser(query: UserQuery): Promise<IUser[]> {
+    return this._matchUsers(query);
+  }
+
+  async getTemplate(query: TemplateQuery): Promise<ITemplate[]> {
+    return this._matchTemplates(query);
+  }
+
+  // Will be fixed in a while to use JSONResponse
+  async removeUser(query: UserQuery): Promise<JSONResponse<Number>> {
+    let removeCount: number = 0;
+    await this._matchUsers(query).then(users => {
+      users.forEach(user => {
+        removeCount += 1;
+        this.users.delete(user._id!);
+      });
+    });
+    return Promise.resolve({ success: true, result: removeCount });
+  }
+  async removeTemplate(query: TemplateQuery): Promise<JSONResponse<Number>> {
+    let removeCount: number = 0;
+    await this._matchTemplates(query).then(templates => {
+      templates.forEach(template => {
+        removeCount += 1;
+        this.templates.delete(template._id!);
+      });
+    });
+    return Promise.resolve({ success: true, result: removeCount });
+  }
+
+  protected async _matchUsers(query: UserQuery): Promise<IUser[]> {
+    let res: IUser[] = new Array();
+    this.users.forEach(user => {
+      if (this._matchUser(query, user)) {
+        res.push(this._clone(user));
+      }
+    });
+    return Promise.resolve(res);
+  }
+
+  protected async _matchTemplates(query: TemplateQuery): Promise<ITemplate[]> {
+    let res: ITemplate[] = new Array();
+    this.templates.forEach(template => {
+      if (this._matchTemplate(query, template)) {
+        res.push(this._clone(template));
+      }
+    });
+    return Promise.resolve(res);
+  }
+
+  protected _clone<T>(obj: T): T {
+    let cloned: T = JSON.parse(JSON.stringify(obj));
+    return cloned;
+  }
+
+  protected async _insert<T extends ITemplate | IUser>(doc: T, collection: Map<String, T>): Promise<void> {
+    let docToInsert: T = this._clone(doc);
+    this._checkID(docToInsert);
+    if (!collection.has(docToInsert._id!)) {
+      collection.set(docToInsert._id!, docToInsert);
+    } else {
+      logger.error("Object with id: %s already exists. Insertion failed", doc._id!);
+    }
+  }
+
+  // Makes sure that id of the object is set
+  protected _checkID<T extends ITemplate | IUser>(doc: T) {
+    if (!doc._id) {
+      doc._id = uuidv4();
+    }
+  }
+
+  protected _setTimestamps(doc: ITemplate): void {
+    doc.createdAt = new Date(Date.now());
+  }
+
+  protected _matchUser(query: UserQuery, user: IUser): boolean {
+    if (query.userID && !(query.userID === user._id)) {
+      return false;
+    }
+    if (query.email && !(query.email === user.email)) {
+      return false;
+    }
+
+    if (query.orgID) {
+      for (let org of query.orgID) {
+        if (!user.org.includes(org)) {
+          return false;
+        }
+      }
```

Comment:
it seems like there are a lot of repetition of this code.
You can write a function for that and call it in other places, as all the types are string and string[] I guess?